src/dune_rules/lib.ml

```diff
@@ -472,29 +507,204 @@ let is_local t =
+         | Gt -> Error.too_many_arguments ~loc:arg.loc arg.arg.info)
+      | ((_, None) as keep) :: existing, [] -> go (keep :: acc) existing []
+    in
+    let* t_arguments = parameterized_arguments t in
+    let+ arguments = go [] t_arguments new_arguments in
+    let arguments = List.map ~f:snd arguments in
+    { t with arguments }
+  ;;
+
+  let make_argument (loc, arg) =
     let open Resolve.O in
-    let* lib = lib in
-    let* required_parameters = lib.parameters in
+    let* arg = arg in
+    let* param =
+      match arg.implements with
+      | Some param -> param
+      | None -> Error.missing_implements ~loc arg.info
+    in
+    let+ param_name = resolve_main_module_name param
+    and+ arg_name = resolve_main_module_name arg in
+    ( param
+    , { arg
+      ; param_name = Option.value_exn param_name
+      ; arg_name = Option.value_exn arg_name
+      ; loc
+      } )
+  ;;
+
+  let make_arguments arguments =
+    let open Resolve.O in
+    let+ arguments = Resolve.List.map arguments ~f:make_argument in
+    List.sort arguments ~compare:(fun (param, _) (param', _) -> compare param param')
+  ;;
+
+  let instantiate ~loc lib args ~parent_parameters =
+    let open Resolve.O in
+    let* lib = lib
+    and* args = make_arguments args in
+    let* lib = apply_arguments ~ignore_extra:false lib args in
     let+ () =
-      Resolve.List.iter required_parameters ~f:(function
-        | param when not (List.exists parameters ~f:(equal param)) ->
+      let* all_args = parameterized_arguments lib in
+      Resolve.List.iter all_args ~f:(function
+        | param, None when not (List.exists parent_parameters ~f:(equal param)) ->
           Error.missing_parameter ~loc param.info
         | _ -> Resolve.return ())
     in
     lib
   ;;
+
+  let complement_arguments ~parent dep =
+    let open Resolve.O in
+    let* parent_arguments = parameterized_arguments parent in
+    let parent_arguments =
+      List.filter_map parent_arguments ~f:(fun (param, opt_arg) ->
+        Option.map opt_arg ~f:(fun arg -> param, arg))
+    in
+    let* arguments =
+      Resolve.List.map dep.arguments ~f:(fun opt_arg ->
+        match opt_arg with
+        | None -> Resolve.return None
+        | Some argument ->
+          let+ arg = apply_arguments ~ignore_extra:true argument.arg parent_arguments in
+          Some { argument with arg })
+    in
+    apply_arguments ~ignore_extra:true { dep with arguments } parent_arguments
+  ;;
+
+  let remove_arguments lib = { lib with parameters = Resolve.return []; arguments = [] }
+
+  let requires lib =
+    let open Resolve.O in
+    let* (deps : lib list) = lib.requires in
+    let* deps =
+      Resolve.List.map deps ~f:(fun dep -> complement_arguments ~parent:lib dep)
+    in
+    let lib_arguments =
+      List.filter_map lib.arguments ~f:(function
+        | None -> None
+        | Some arg -> Some arg.arg)
+    in
+    let deps = List.rev_append lib_arguments deps in
```

Comment:
Is this `rev` necessary? If it's re-ordering something provided by the user, I'd suggest not to do it.

Reply:
No it's not necessary, I've removed it in favor of `@` since I didn't want to imply that the order matters and `lib_arguments` is expected to be small anyway. At this point the list of arguments to instantiate the library have already been reordered to match the ordering of the parameters.